src/GafferScene/MeshSegments.cpp

```diff
@@ -0,0 +1,277 @@
+			{
+				int vNext = segments[v];
+				segments[v] = minVertex;
+				if( vNext == v )
+				{
+					break;
+				}
+				v = vNext;
+			}
+		}
+
+		// Advance to next face
+		vertexIdsIndex += f;
+	}
+
+	// We now have all faces considered, and have the proprety that every vertex points to a vertex less than
+	// itself in the segment unless it is lowest in the segment.  This means we can now just process all
+	// vertices, starting from the lowest.  If a vertex points to itself, it marks a new segment, otherwise
+	// it can just take the segment index from the vertex it points to ( which is guaranteed to have already
+	// been processed, since we process in order ).
+	int numSegments = 0;
+	for( int i = 0; i < numIndexed; i++ )
+	{
+		if( segments[i] == i )
+		{
+			numSegments++;
+			segments[i] = numSegments - 1;
+		}
+		else
+		{
+			segments[i] = segments[ segments[i] ];
+		}
+	}
+
+	uniformSegments.clear();
+	uniformSegments.reserve( verticesPerFace.size() );
+
+	// Convert from whatever "vertices" we are segmenting ( which may actually be UVs or anything else
+	// that is indexed ) to uniform ( one value per face ).  We do this just by reading one vertex from
+	// each face.
+	vertexIdsIndex = 0;
+	for( int f : verticesPerFace )
+	{
+		uniformSegments.push_back( segments[ indices[ vertexIdsIndex ] ] );
+		vertexIdsIndex += f;
+	}
+}
+
+} // namespace
+
+size_t MeshSegments::g_firstPlugIndex = 0;
+
+GAFFER_NODE_DEFINE_TYPE( MeshSegments );
+
+MeshSegments::MeshSegments( const std::string &name )
+	:	ObjectProcessor( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	addChild( new StringPlug( "segmentsPrimVar", Plug::In, "segments" ) );
+	addChild( new StringPlug( "indexPrimVar", Plug::In, "P" ) );
+}
+
+MeshSegments::~MeshSegments()
+{
+}
+
+Gaffer::StringPlug *MeshSegments::segmentsPrimVarPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+const Gaffer::StringPlug *MeshSegments::segmentsPrimVarPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex );
+}
+
+Gaffer::StringPlug *MeshSegments::indexPrimVarPlug()
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+const Gaffer::StringPlug *MeshSegments::indexPrimVarPlug() const
+{
+	return getChild<StringPlug>( g_firstPlugIndex + 1 );
+}
+
+bool MeshSegments::affectsProcessedObject( const Gaffer::Plug *input ) const
+{
+	return
+		ObjectProcessor::affectsProcessedObject( input ) ||
+		input == segmentsPrimVarPlug() ||
+		input == indexPrimVarPlug()
+	;
+}
+
+void MeshSegments::hashProcessedObject( const ScenePath &path, const Gaffer::Context *context, IECore::MurmurHash &h ) const
+{
+	ObjectProcessor::hashProcessedObject( path, context, h );
+
+	segmentsPrimVarPlug()->hash( h );
+	indexPrimVarPlug()->hash( h );
+}
+
+IECore::ConstObjectPtr MeshSegments::computeProcessedObject( const ScenePath &path, const Gaffer::Context *context, const IECore::Object *inputObject ) const
+{
+	const std::string &segmentsPrimVar = segmentsPrimVarPlug()->getValue();
```

Comment:
Right, that is misleading - fixed.  I hadn't really though about how often we have to allocate strings in places like this ... not really a concern here, but it would be nice in the general case if there were the ability to just use `getObjectValue<StringData>()->readable()` in the common case where the value doesn't contain any substitutions.

Reply:
Yeah, and we also have a fair few places where we convert immediately to an `InternedString`, or `ScenePath`. Those could also benefit from some optimisation for the case where there are no substitutions.

We haven't switched to the [new C++ ABI](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html) yet, but VFXPlatform does mandate it for 2023. That likely changes the performance characteristics for our strings (it goes from copy-on-write to small-string-optimisation), and might be a good time to investigate the other stuff...